File: e2e/fixtures/ruby/weather-app/lib/weather_app/formatter.rb
require "json"

module WeatherApp
  class Formatter
    FORMATS = %i[text json].freeze

    def initialize(format: :text)
      raise ArgumentError, "unknown format #{format}" unless FORMATS.include?(format)
      @format = format
    end

    def render(forecast)
      case @format
      when :text then render_text(forecast)
      when :json then JSON.generate(forecast.to_h)
      end
    end

    private

    def render_text(forecast)
      [
        "City:        #{forecast.city}",
        "Condition:   #{forecast.condition}",
        "Temperature: #{format('%.1f', forecast.temperature_c)} C / " \
          "#{format('%.1f', forecast.temperature_f)} F",
        "Humidity:    #{forecast.humidity}%"
      ].join("\n")
    end
  end
end

File: e2e/fixtures/java/weather-app/src/main/java/com/example/weather/formatter/Formatter.java
package com.example.weather.formatter;

import com.example.weather.model.Forecast;

import java.util.Locale;

public class Formatter {
    public enum Format { TEXT, JSON }

    private final Format format;

    public Formatter(Format format) { this.format = format; }

    public String render(Forecast f) {
        return switch (format) {
            case TEXT -> renderText(f);
            case JSON -> renderJson(f);
        };
    }

    private String renderText(Forecast f) {
        return String.format(Locale.ROOT,
            "City:        %s%nCondition:   %s%nTemperature: %.1f C / %.1f F%nHumidity:    %d%%",
            f.city(), f.condition(), f.temperatureC(), f.temperatureF(), f.humidity());
    }

    private String renderJson(Forecast f) {
        return String.format(Locale.ROOT,
            "{\"city\":\"%s\",\"temperature_c\":%.1f,\"temperature_f\":%.1f,\"condition\":\"%s\",\"humidity\":%d}",
            f.city(), f.temperatureC(), f.temperatureF(), f.condition(), f.humidity());
    }
}
